packages/diagrams/lib/src/basic_shapes.dart

```diff
@@ -0,0 +1,1050 @@
+}) {
+  if (control) {
+    final Rect rect = Rect.fromCircle(center: offset, radius: 4.0);
+    canvas.drawRect(rect, Paint()..color = color);
+  } else {
+    canvas.drawCircle(
+      offset,
+      4,
+      Paint()..color = color,
+    );
+  }
+  if (label != null) {
+    paintLabel(
+      canvas,
+      label,
+      offset: offset,
+      padding: padding,
+      alignment: alignment,
+      style: _kLabelStyle,
+    );
+  }
+}
+
+void _paintXYPlane(
+  Canvas canvas, {
+  int width = 10,
+  int height = 6,
+  Color color = _kForegroundColor,
+}) {
+  final double rightEdge = _kGridSize * width;
+  final double bottomEdge = _kGridSize * height;
+  const double arrowNudge = 8.0;
+  final double rightArrow = rightEdge - arrowNudge;
+  final double bottomArrow = bottomEdge - arrowNudge;
+
+  final Paint paint = Paint()
+    ..color = color
+    ..style = PaintingStyle.stroke
+    ..strokeWidth = 3.5;
+
+  // Draw main lines going to the right and down
+
+  canvas.drawPath(
+    Path()
+      ..moveTo(rightArrow, 0)
+      ..lineTo(0, 0)
+      ..lineTo(0, bottomArrow),
+    paint,
+  );
+
+  // Draw arrows
+
+  paint
+    ..color = color
+    ..style = PaintingStyle.fill;
+  canvas.drawPath(
+    Path()
+      ..moveTo(rightArrow, -5)
+      ..lineTo(rightArrow + 10, 0)
+      ..lineTo(rightArrow, 5),
+    paint,
+  );
+  canvas.drawPath(
+    Path()
+      ..moveTo(-5, bottomArrow)
+      ..lineTo(0, bottomArrow + 10)
+      ..lineTo(5, bottomArrow),
+    paint,
+  );
+
+  // Draw labels
+
+  paintLabel(
+    canvas,
+    '0,0',
+    offset: const Offset(-4, -4),
+    alignment: Alignment.topLeft,
+    style: _kLabelStyle,
+  );
+
+  paintLabel(
+    canvas,
+    '+x',
+    offset: Offset(rightArrow, -8),
+    alignment: Alignment.topCenter,
+    style: _kLabelStyle,
+  );
+
+  paintLabel(
+    canvas,
+    '+y',
+    offset: Offset(-8, bottomArrow),
+    alignment: Alignment.centerLeft,
+    style: _kLabelStyle,
+  );
+}
+
+class LineDiagramPainter extends CustomPainter {
+  @override
+  void paint(Canvas canvas, Size size) {
+    canvas.save();
+    canvas.translate(56.0, 48.0);
+
+    _paintXYPlane(canvas);
+
+    final Paint paint = Paint()
+      ..strokeWidth = 5.0
+      ..style = PaintingStyle.stroke
+      ..color = _kForegroundColor;
+
+    final Offset start = const Offset(2, 4) * _kGridSize;
+    final Offset end = const Offset(8, 2) * _kGridSize;
+
+    canvas.drawLine(
+      start,
+      end,
+      paint,
+    );
+
+    paintLabel(
+      canvas,
+      'p1',
+      offset: start,
+      alignment: Alignment.topCenter,
+      padding: _kLabelPadding,
+      style: _kLabelStyle,
+    );
+
+    paintLabel(
+      canvas,
+      'p2',
+      offset: end,
+      alignment: Alignment.topCenter,
+      padding: _kLabelPadding,
+      style: _kLabelStyle,
+    );
+
+    canvas.restore();
+  }
+
+  @override
+  bool shouldRepaint(LineDiagramPainter oldDelegate) => true;
+}
+
+class RectCtorDiagramPainter extends CustomPainter {
```

Comment:
```suggestion
class RectConstructorDiagramPainter extends CustomPainter {
```

We try not to use abbreviations in names.